File: src/net/KNet/Specific/Connect/Transforms/Predicates/KNetPredicate.cs
/*
*  Copyright (c) 2021-2026 MASES s.r.l.
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*  http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
*
*  Refer to LICENSE for more information.
*/

using Java.Lang;
using Java.Util;
using MASES.JCOBridge.C2JBridge;
using MASES.JCOBridge.C2JBridge.JVMInterop;
using MASES.JNet.Specific.Extensions;
using Org.Apache.Kafka.Common.Config;
using Org.Apache.Kafka.Connect.Connector;
using Org.Apache.Kafka.Connect.Errors;
using Org.Apache.Kafka.Connect.Sink;
using Org.Apache.Kafka.Connect.Source;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace MASES.KNet.Connect.Transforms.Predicates
{
    /// <summary>
    /// .NET interface for <see cref="IVersion"/>
    /// </summary>
    public interface IVersion
    {
        /// <summary>
        /// Returns version string
        /// </summary>
        string Version();
    }
    /// <summary>
    /// .NET interface for <see cref="Connector"/>
    /// </summary>
    public interface IPredicate : IVersion
    {
        /// <inheritdoc cref="Org.Apache.Kafka.Connect.Transforms.Predicates.Predicate.Test(ConnectRecord)"/>
        bool Test(ConnectRecord record);
        /// <inheritdoc cref="Org.Apache.Kafka.Common.Configurable.Configure(Map{Java.Lang.String, object})"/>
        void Configure(Map<Java.Lang.String, object> props);
        /// <inheritdoc cref="Org.Apache.Kafka.Connect.Transforms.Predicates.Predicate.Close"/>
        void Close();
        /// <inheritdoc cref="Org.Apache.Kafka.Connect.Transforms.Predicates.Predicate.Config"/>
        ConfigDef Config();
        /// <summary>
        /// Returns a formatted string used from <see cref="Org.Apache.Kafka.Connect.Transforms.Predicates.Predicate"/>
        /// </summary>
        /// <returns>The formatted string</returns>
        string ToStringPredicate();
    }

    /// <summary>
    /// Specific implementation of <see cref="IConnector"/> to support KNet Connect SDK
    /// </summary>
    public interface IKNetPredicate : IPredicate
    {
        /// <summary>
        /// Implements the behavior of <see cref="IPredicate.Test(ConnectRecord)"/> for <paramref name="record"/>
        /// </summary>
        /// <param name="record">The <see cref="SourceRecord"/> to test</param>
        /// <returns>Follow specifications of <see cref="Org.Apache.Kafka.Connect.Transforms.Predicates.Predicate.Test(ConnectRecord)"/></returns>
        /// <remarks>If the method is overridden never invoke the base method othrwise an exception is raised within the JVM.</remarks>
        bool Test(SourceRecord record);
        /// <summary>
        /// Implements the behavior of <see cref="IPredicate.Test(ConnectRecord)"/> for <paramref name="record"/>
        /// </summary>
        /// <param name="record">The <see cref="SinkRecord"/> to test</param>
        /// <returns>Follow specifications of <see cref="Org.Apache.Kafka.Connect.Transforms.Predicates.Predicate.Test(ConnectRecord)"/></returns>
        /// <remarks>If the method is overridden never invoke the base method othrwise an exception is raised within the JVM.</remarks>
        bool Test(SinkRecord record);
        /// <summary>
        /// Implement the method to execute the start action
        /// </summary>
        /// <param name="configuration">The <see cref="IKNetConfigurationFromMap"/> to access the properties returned from Apache Kafka Connect framework: the <see cref="IKNetCommon.Properties"/> contains the same info from configuration file.</param>
        void Configure(IKNetConfigurationFromMap configuration);
    }
    /// <summary>
    /// The generic class which is the base of all predicates in .NET
    /// </summary>
    public abstract class KNetPredicate : KNetCommon, IKNetPredicate
    {
        /// <summary>
        /// Set the <see cref="ReflectedRemoteObjectClassName"/> of the connector to a fixed value
        /// </summary>
        protected sealed override string ReflectedRemoteObjectClassName => "KNetPredicate";

        /// <summary>
        /// Public method used from Java to trigger <see cref="Test(ConnectRecord)"/>
        /// </summary>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        public bool TestInternal()
        {
            try
            {
                using var record = DataToExchange<ConnectRecord>();
                return Test(record);
            }
            catch (System.Exception e)
            {
                LogError($"TestInternal failed with {e}");
                throw;
            }
        }

        /// <inheritdoc cref="IPredicate.Test(ConnectRecord)"/>
        public virtual bool Test(ConnectRecord record)
        {
            if (record is null) return false;

            if (record.IsInstanceOf<SourceRecord>())
            {
                return Test(record.CastDirectAndDetach<SourceRecord>());
            }
            else if (record.IsInstanceOf<SinkRecord>())
            {
                return Test(record.CastDirectAndDetach<SinkRecord>());
            }
            else JVMBridgeException.Throw<ConnectException>($"Cannot manage directly the input, override the method {nameof(Test)} with generic {nameof(ConnectRecord)} parameter.");
            return false;
        }

        /// <inheritdoc cref="IKNetPredicate.Test(SourceRecord)"/>
        public virtual bool Test(SourceRecord record)
        {
            JVMBridgeException.Throw<ConnectException>($"Not implemented for {nameof(SourceRecord)}");
            return false;
        }

        /// <inheritdoc cref="IKNetPredicate.Test(SinkRecord)"/>
        public virtual bool Test(SinkRecord record)
        {
            JVMBridgeException.Throw<ConnectException>($"Not implemented for {nameof(SinkRecord)}");
            return false;
        }

        /// <summary>
        /// Public method used from Java to trigger <see cref="ToStringInternal"/>
        /// </summary>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        public string ToStringInternal()
        {
            try
            {
                return ToStringPredicate();
            }
            catch (System.Exception e)
            {
                LogError($"ToStringInternal failed with {e}");
                throw;
            }
        }

        /// <inheritdoc cref="IPredicate.ToStringPredicate"/>
        public virtual string ToStringPredicate() { return null; }

        /// <summary>
        /// Public method used from Java to trigger <see cref="Configure(Map{Java.Lang.String, object})"/>
        /// </summary>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        public void ConfigureInternal()
        {
            try
            {
                Map<Java.Lang.String, object> props = DataToExchange<Map<Java.Lang.String, object>>();
                Configure(props);
                Properties = new KNetConfigurationFromMap(props);
                Configure(Properties);
            }
            catch (System.Exception e)
            {
                LogError($"ConfigureInternal failed with {e}");
                throw;
            }
        }
        /// <summary>
        /// Not implemented
        /// </summary>
        /// <exception cref="NotImplementedException">Local version with a different signature</exception>
        public virtual void Configure(Map<Java.Lang.String, object> props)
        {

        }

        /// <inheritdoc cref="IKNetPredicate.Configure(IKNetConfigurationFromMap)"/>
        public abstract void Configure(IKNetConfigurationFromMap configuration);

        /// <summary>
        /// Public method used from Java to trigger <see cref="Close"/>
        /// </summary>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        public void CloseInternal()
        {
            try
            {
                Close();
            }
            catch (System.Exception e)
            {
                LogError($"CloseInternal failed with {e}");
            }

            try
            {
                Unregister();
            }
            catch { }
        }
        /// <summary>
        /// Implement the method to execute the close action
        /// </summary>
        public virtual void Close() { }
        /// <summary>
        /// Not implemented
        /// </summary>
        /// <exception cref="NotImplementedException">Invoked in Java before any initialization</exception>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        public ConfigDef Config() => throw new NotImplementedException("Invoked in Java before any initialization.");
        /// <summary>
        /// Not implemented
        /// </summary>
        /// <exception cref="NotImplementedException">Invoked in Java before any initialization</exception>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        public string Version() => throw new NotImplementedException("Invoked in Java before any initialization.");
    }
}

File: src/jvm/knet/src/main/java/org/mases/knet/developed/connect/transforms/predicates/KNetPredicate.java
/*
 *  Copyright (c) 2021-2026 MASES s.r.l.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Refer to LICENSE for more information.
 */

package org.mases.knet.developed.connect.transforms.predicates;

import org.apache.kafka.common.config.AbstractConfig;
import org.apache.kafka.common.config.ConfigDef;
import org.apache.kafka.connect.errors.ConnectException;
import org.apache.kafka.connect.components.Versioned;
import org.apache.kafka.connect.connector.ConnectRecord;
import org.apache.kafka.connect.transforms.predicates.Predicate;
import org.mases.knet.developed.connect.*;
import org.mases.jcobridge.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Map;

import org.mases.knet.developed.connect.KNetConnectInitializer;

public class KNetPredicate<R extends ConnectRecord<R>> implements Predicate<R>, Versioned, KNetConnectLogging, KNetConnectInitializer, KNetConnectDataExchange {
    private static final Logger log = LoggerFactory.getLogger(KNetPredicate.class);

    private static final String registrationName = "KNetPredicate";

    public static final ConfigDef CONFIG_DEF = new ConfigDef();

    long predicateId = 0;

    String indexedRegistrationName;

    JCObject predicateObject;

    Object dataToExchange = null;

    public KNetPredicate() throws JCException, IOException {
        super();
        if (!JCOBridge.isCLRHostingProcess()) {
            KNetConnectProxy.initAndGetConnectProxy();
        } else {
            throw new ConnectException("KNetPredicate is not supported from a CLR Hosting process.");
        }
    }

    public String getAssemblyLocation() {
        return null;
    }

    public String getClassName() {
        return null;
    }

    public Object getDataToExchange() {
        return dataToExchange;
    }

    public void setDataToExchange(Object dte) {
        dataToExchange = dte;
    }

    @Override
    public boolean test(R record) {
        log.debug("Invoking test");
        try {
            if (predicateObject != null) {
                try {
                    dataToExchange = record;
                    return (boolean) predicateObject.Invoke("TestInternal");
                } finally {
                    dataToExchange = null;
                }
            } else {
                log.warn("Cannot execute \"test\" since remote object is missing");
                return false;
            }
        } catch (JCException jcne) {
            log.error("Failed Invoke of \"test\"", jcne);
            throw new ConnectException("Failed Invoke of \"test\"", jcne);
        }
    }

    @Override
    public String version() {
        log.debug("Invoking version");
        try {
            if (predicateObject != null) {
                return (String) predicateObject.Invoke("VersionInternal");
            }
        } catch (JCException jcne) {
            log.error("Failed Invoke of \"version\"", jcne);
        }
        return "NOT AVAILABLE";
    }

    @Override
    public ConfigDef config() {
        return CONFIG_DEF;
    }

    @Override
    public void close() {
        log.debug("Invoking close");
        try {
            try {
                if (predicateObject != null) {
                    predicateObject.Invoke("CloseInternal");
                }
            } finally {
                if (!JCOBridge.isCLRHostingProcess()) {
                    JCOBridge.UnregisterJVMGlobal(indexedRegistrationName);
                }
            }
        } catch (JCException jcne) {
            log.error("Failed Invoke of \"close\"", jcne);
        }
    }

    @Override
    public void configure(Map<String, ?> configs) {
        log.debug("Invoking configure");
        try {
            predicateId = KNetConnectProxy.getNewConnectorId();
            if (JCOBridge.isCLRHostingProcess()) {
                throw new ConnectException("KNetPredicate is not supported from a CLR Hosting process.");
            } else {
                indexedRegistrationName = String.format("%s_%d", registrationName, predicateId);
                log.info("Preparing KNetTransform with name {}", indexedRegistrationName);
                if (!KNetConnectProxy.initializePredicate(this, configs, indexedRegistrationName)) {
                    log.error("Failed Invoke of \"initializePredicate\"");
                    throw new ConnectException("Failed Invoke of \"initializePredicate\"");
                }
                JCOBridge.RegisterJVMGlobal(indexedRegistrationName, this);
                log.info("RegisterJVMGlobal done for {}", indexedRegistrationName);
                predicateObject = KNetConnectProxy.getPredicate(indexedRegistrationName);
            }
            if (predicateObject != null) {
                try {
                    AbstractConfig config = new AbstractConfig(config(), configs);
                    dataToExchange = config.values();
                    predicateObject.Invoke("ConfigureInternal");
                } finally {
                    dataToExchange = null;
                }
            }
        } catch (JCException | IOException jcne) {
            log.error("Failed Invoke of \"configure\"", jcne);
            throw new ConnectException("Failed Invoke of \"configure\"", jcne);
        }
    }

    @Override
    public String toString() {
        log.debug("Invoking toString");
        String result = null;
        try {
            if (predicateObject != null) {
                result = (String) predicateObject.Invoke("ToStringInternal");
            }
        } catch (JCException jcne) {
            log.error("Failed Invoke of \"toString\"", jcne);
        }
        return result != null ? result : super.toString();
    }

    @Override
    public String getName() { return log.getName(); }

    @Override
    public boolean isTraceEnabled() {
        return log.isTraceEnabled();
    }

    @Override
    public void trace(String var1) { log.trace(var1); }

    @Override
    public void trace(String var1, Throwable var2) {
        log.trace(var1, var2);
    }

    @Override
    public void trace(String var1, Object... var2) {
        log.trace(var1, var2);
    }

    @Override
    public boolean isDebugEnabled() {
        return log.isDebugEnabled();
    }

    @Override
    public void debug(String var1) {
        log.debug(var1);
    }

    @Override
    public void debug(String var1, Throwable var2) {
        log.debug(var1, var2);
    }

    @Override
    public void debug(String var1, Object... var2) {
        log.trace(var1, var2);
    }

    @Override
    public boolean isInfoEnabled() {
        return log.isInfoEnabled();
    }

    @Override
    public void info(String var1) {
        log.info(var1);
    }

    @Override
    public void info(String var1, Throwable var2) {
        log.info(var1, var2);
    }

    @Override
    public void info(String var1, Object... var2) {
        log.trace(var1, var2);
    }

    @Override
    public boolean isWarnEnabled() {
        return log.isWarnEnabled();
    }

    @Override
    public void warn(String var1) {
        log.warn(var1);
    }

    @Override
    public void warn(String var1, Throwable var2) {
        log.warn(var1, var2);
    }

    @Override
    public void warn(String var1, Object... var2) { log.trace(var1, var2); }

    @Override
    public boolean isErrorEnabled() {
        return log.isErrorEnabled();
    }

    @Override
    public void error(String var1) {
        log.error(var1);
    }

    @Override
    public void error(String var1, Throwable var2) {
        log.error(var1, var2);
    }

    @Override
    public void error(String var1, Object... var2) {
        log.trace(var1, var2);
    }
}
